### ISSUE TITLE
The project of E5: How to fine-tune on the E5 model?

### ISSUE THREAD

**Issue by EasyLuck** [microsoft/unilm#1066]
Dear author:
The following informations is about the paper and project of E5: [Text Embeddings by Weakly-Supervised Contrastive Pre-training](https://arxiv.org/pdf/2212.03533.pdf). Liang Wang, Nan Yang, Xiaolong Huang, Binxing Jiao, Linjun Yang, Daxin Jiang, Rangan Majumder, Furu Wei, arXiv 2022
I want to fine-tune on my dataset. However, I do not know how to do it, there is no more information on the main site of GitHub: https://github.com/microsoft/unilm/tree/master/e5.  
1. I don't know how to prepare my data. Same as what is  the data structure should I design.
2. What the tools should I use?
Can you provide more information or train_script / finetune_script on the GitHub .
Thank you very much.

**Comment by intfloat** [comment]
Hi,

If you'd like to finetune from our checkpoint, you can refer to our codebase of [SimLM](https://github.com/microsoft/unilm/tree/master/simlm), which contains step-by-step guide to train dense retrievers. Simply change the model initialization and format your dataset accordingly should work. Please check out https://huggingface.co/datasets/intfloat/simlm-msmarco for data format.

Thanks,
Liang

**Comment by kishorenc** [comment]
@intfloat 

In the SimLM repo, I see the following 2 training scripts:

1. Train a biencoder retriever with BM25 hard negatives (pre-training)
2. Train a biencoder retriever with knowledge distillation (fine-tuning)

Both of them are trained from the `intfloat/simlm-base-msmarco` pre-trained model. 

If I wish to fine-tune the E5 model (e.g `e5-small`) on my dataset, which trainer should I use? Should I also furst pre-train and then fine-tune? I will appreciate if you can explain the difference. Thank you in advance.

**Comment by intfloat** [comment]
Knowledge distillation from a cross-encoder model leads to better performance, but it requires training a cross-encoder reranker as a first step. If you want to have a fast proof-of-concept experiment, I recommend initializing from E5 models and training with BM25 hard negatives on your own dataset.

**Comment by deborshigoswami95** [comment]
Hi, thanks so much for your response to the original question. It helped a lot in understanding the quicker method for fine-tuning. I am also interested in the more involved `knowledge distillation from cross-encoder` method. I have a few questions

1. In your paper, I didn't see a time estimate for the fine-tuning. Would you mind giving an intuition for how long it would take to train a cross-encoder reranker & then 3 epochs of fine-tuning?
2. were the 8 GPUs used for data parallelism? Or did you also have to parallelize the model a la Parallelformers?
3. How much quality labeled data would you recommend for fine-tuning to even be worth it?

Thank you very much in advance and I appreciate the insight already provided!

**Comment by intfloat** [comment]
For your questions:

1. I do not remember the exact numbers, but training cross-encoder reranker and bi-encoder fine-tuning are pretty fast, they should take less than 12 hours each with 8 V100 GPUs. If you have A100, the time will be even shorter. 

2. Yes, we use 8 V100 GPUs with pytorch distributed data parallelism (DDP). No model parallelism is used since the models are not that large anyway.

3. E5 models are fine-tuned with about 900k labeled examples. I would expect knowledge distillation from cross-encoder to have performance gains even if you only have several thousands of examples.

**Comment by deborshigoswami95** [comment]
Really appreciate the speedy response! That clears up a lot, thank you :)

**Comment by LazyFyh** [comment]
Hi,
I want to evaluate E5 model on trecdl19 and trecdl20, but MTEB does not seem to provide a special interface for these two datasets, can I put the E5 checkpoint directly on the evaluation code in the codebase of SimLM? Does this operation will result in any loss of precision? Or is there a better way to evaluate?

Thank you very much!


> Hi,
> 
> If you'd like to finetune from our checkpoint, you can refer to our codebase of [SimLM](https://github.com/microsoft/unilm/tree/master/simlm), which contains step-by-step guide to train dense retrievers. Simply change the model initialization and format your dataset accordingly should work. Please check out https://huggingface.co/datasets/intfloat/simlm-msmarco for data format.
> 
> Thanks, Liang

**Comment by intfloat** [comment]
No, the input format is different. E5 models do not use token_type_ids and have additional prefix "query: " and "passage: ".

You can create a custom MTEB task following https://github.com/embeddings-benchmark/mteb/tree/main#evaluating-on-a-custom-task

It is also possible to change the data collators in SimLM codebase to make the input format consistent.

**Comment by LazyFyh** [comment]
> No, the input format is different. E5 models do not use token_type_ids and have additional prefix "query: " and "passage: ".
> 
> You can create a custom MTEB task following https://github.com/embeddings-benchmark/mteb/tree/main#evaluating-on-a-custom-task
> 
> It is also possible to change the data collators in SimLM codebase to make the input format consistent.

Thanks for your recommendation! I'll try it out.

**Comment by wangyiran33** [comment]
> 
Hi, Thanks for previous answer. I have a follow-up question:
Would initializing from E5 models and training with E5 hard negatives on my own dataset works better than BM25 hard negatives since Retriever2 performs better than Retriever1 according to the SIMLM paper.
Thank you very much in advance!

**Comment by intfloat** [comment]
> > 
> 
> Hi, Thanks for previous answer. I have a follow-up question: Would initializing from E5 models and training with E5 hard negatives on my own dataset works better than BM25 hard negatives since Retriever2 performs better than Retriever1 according to the SIMLM paper. Thank you very much in advance!

 Yes, I would guess so.

**Comment by SnehaMadiri** [comment]
Hello, regarding the scoring of positive and negative documents in your training dataset: Can you explain the method used to calculate these scores from ordinary positive and negative text responses?

**Comment by hfawaz** [comment]
What is the approximate dataset size that I would need for fine-tuning e5 on my own dataset ?

**Comment by SKT799** [comment]
@intfloat Hi, Urgent Help needed: 
In the Original E5 paper, NQ+MS MARCO are distilled via simLM but NLI is used simply via infoNCE? You have not implemented Knowledge distillation for the NLI?

Also  if knowledge distillation is not done for NLI then did you guys sequentially finetuned the e5-PT-base? like doing firstly with NLI then NQ+MS MARCO?

I will be grateful if you can answer
